core/src/main/java/smile/classification/DecisionTree.java

```diff
@@ -981,21 +1001,28 @@ public DecisionTree(Attribute[] attributes, double[][] x, int[] y, int maxNodes,
         root = new Node(Math.whichMax(count), posteriori);
         
         TrainNode trainRoot = new TrainNode(root, x, y, samples, 0, originalOrder.length);
-        // Now add splits to the tree until max tree size is reached
-        if (trainRoot.findBestSplit()) {
-            nextSplits.add(trainRoot);
-        }
-
-        // Pop best leaf from priority queue, split it, and push
-        // children nodes into the queue if possible.
-        for (int leaves = 1; leaves < this.maxNodes; leaves++) {
-            // parent is the leaf to split
-            TrainNode node = nextSplits.poll();
-            if (node == null) {
-                break;
+        if(maxNodes == Integer.MAX_VALUE) {// depth-first split
+            if (trainRoot.findBestSplit()) {
+                trainRoot.split(null);
             }
+        } else {// best-first split
+            // Priority queue for best-first tree growing.
+            PriorityQueue<TrainNode> nextSplits = new PriorityQueue<>();
 
-            node.split(nextSplits); // Split the parent node into two children nodes
+            // Now add splits to the tree until max tree size is reached
+            if (trainRoot.findBestSplit()) {
+                nextSplits.add(trainRoot);
+            }
+            // Pop best leaf from priority queue, split it, and push
+            // children nodes into the queue if possible.
+            for (int leaves = 1; leaves < this.maxNodes; leaves++) {
+                // parent is the leaf to split
+                TrainNode node = nextSplits.poll();
+                if (node == null) {
+                    break;
+                }
+                node.split(nextSplits); // Split the parent node into two children nodes
```

Comment:
I think this should be
if (!node.split(nextSplits)) { leaves--; }
because if the node was split into two leaves with the same label and the split was aborted because of that, then the split shouldn't count towards the tree's leaf limit.

Reply:
You're right. 

The same problem exists in the current master as seen in 
https://github.com/haifengl/smile/blob/master/core/src/main/java/smile/classification/DecisionTree.java#L997

Reply:
True - but if I've followed the logic correctly, I think that in the current master node.split() will never return false; after your change it will return false in some cases.

Why I think it won't return false at the moment:
TrainNode.split() can only be called if node.splitFeature >= 0. It will return false only if applying the split specified by (node.splitFeature, node.splitValue) results in either the true branch or the false branch having a size < nodeSize. If both the true and false branches are >= nodeSize, then it must (currently) return true. So in order for it to return false, it needs to be holding a node having a split criterion that produces a too-small branch on one side or the other. That split criterion needs to have been generated by one of the calls to findBestSplit(n, ...) - but https://github.com/haifengl/smile/blob/master/core/src/main/java/smile/classification/DecisionTree.java#L595 and https://github.com/haifengl/smile/blob/master/core/src/main/java/smile/classification/DecisionTree.java#L633 won't approve such a criterion; in order for splitFeature/splitValue to get set, the code has to get past one or the other of those checks, meaning that both the true and false branches will be large enough... so split() can't currently return false.

Aside: those two checks in turn seem to imply that the checks at https://github.com/haifengl/smile/blob/master/core/src/main/java/smile/classification/DecisionTree.java#L507, https://github.com/haifengl/smile/blob/master/core/src/main/java/smile/classification/DecisionTree.java#L727 and https://github.com/haifengl/smile/blob/master/core/src/main/java/smile/classification/DecisionTree.java#L736 (which shortcut, or refuse to call, findBestSplit() if the candidate's node size <= nodeSize) could be tightened up to check "< 2 * nodeSize" - unless you have at least 2*nodeSize units to spread around, at least one branch is going to be <nodeSize. This would speed up tree construction slightly.

Reply:
I'm not 100% confident [this condition](https://github.com/haifengl/smile/blob/master/core/src/main/java/smile/classification/DecisionTree.java#L711) never be happen while your investigation seems right.

Anyway, returning value of `split()` should be meaningful without assuming it would not return false. 

`leaves--` should be inserted when split() returns false. +1 on this option.
Or, better to throw an exception if unexpected behavior (i.e., return false) happens.

> check "< 2 * nodeSize" - unless you have at least 2*nodeSize units to spread around, at least one branch is going to be <nodeSize.

Yeah, it's better to be ensured.

I'm doing it in [my smile modification](https://github.com/apache/incubator-hivemall/pull/198/commits/294ce93fd4ac2c855da5af74fccfc6213456d5f8#diff-2c14597fd8cb87295b3206710c62eeb4R1148).